### PR TITLE
Allow multiple keys for the `ADMIN_CLIENT_SECRET`

### PR DESCRIPTION
**Pull request by idavidmcdonald**
## What
https://www.pivotaltracker.com/story/show/171352263 - but just does the admin client secret to start small, we can do the other as separate PRs once we are happy with this

## Requires 
https://github.com/alphagov/notifications-credentials/pull/143 to be merged first

## How to test

Pull down this branch
In `app/config.py` change your development config to
`ADMIN_CLIENT_SECRETS = json.loads(os.environ.get('ADMIN_CLIENT_SECRETS', '[]'))`

```
export ADMIN_CLIENT_SECRETS='["key1", "key2"]'
./scripts/run_app.sh
```

Then open the admin app code and change the following
https://github.com/alphagov/notifications-admin/blob/master/app/config.py#L106

to be

`ADMIN_CLIENT_SECRET="key1"`.

Spin up your apps and log into the admin app, this will validate that it can talk using `key1`.

Then quit the admin app, change `ADMIN_CLIENT_SECRET="key2"` and retest again.

### CODE REVIEW

File: app/authentication/auth.py [hunk 1]
```diff
@@ -61,7 +61,22 @@ def requires_admin_auth():
 
     if client == current_app.config.get('ADMIN_CLIENT_USER_NAME'):
         g.service_id = current_app.config.get('ADMIN_CLIENT_USER_NAME')
-        return handle_admin_key(auth_token, current_app.config.get('ADMIN_CLIENT_SECRET'))
+
+        if len(current_app.config.get('ADMIN_CLIENT_SECRETS')):
```
Comment (leohemsted) [inline app/authentication/auth.py]: don't think you need this if check. the for loop will just do nothing if the current app has an empty array

File: app/authentication/auth.py [hunk 2]
```diff
@@ -61,7 +61,22 @@ def requires_admin_auth():
...
+            for secret in current_app.config.get('ADMIN_CLIENT_SECRETS'):
+                try:
+                    decode_jwt_token(auth_token, secret)
+                    return
+                except TokenExpiredError:
+                    raise AuthError("Invalid token: expired, check that your system clock is accurate", 403)
+                except TokenDecodeError:
+                    # TODO: Change this so it doesn't also catch `TokenIssuerError` or `TokenIssuedAtError` exceptions
+                    # (which are children of `TokenDecodeError`) as these should cause an auth error immediately rather
+                    # than continue on to check the next API key
```
Comment (leohemsted) [inline app/authentication/auth.py]: minor, comment suggestion
```suggestion
                    # than continue on to check the next admin client secret
```

File: app/authentication/auth.py (hunk 1 above)
Comment (idavidmcdonald) [reply app/authentication/auth.py]: Fixed

File: app/authentication/auth.py (hunk 2 above)
Comment (idavidmcdonald) [reply app/authentication/auth.py]: Fixed

File: app/config.py [hunk 3]
```diff
@@ -369,7 +369,7 @@ class Development(Config):
     TRANSIENT_UPLOADED_LETTERS = 'development-transient-uploaded-letters'
     LETTER_SANITISE_BUCKET_NAME = 'development-letters-sanitise'
 
-    ADMIN_CLIENT_SECRET = 'dev-notify-secret-key'
+    API_INTERNAL_SECRETS = json.loads(os.environ.get('API_INTERNAL_SECRETS', '[]'))
```
Comment (servingUpAces) [inline app/config.py]: If we aren't going to have a default here can we add a comment in the README (in the section we talk about environment variables), please?

Comment (idavidmcdonald) [reply app/config.py]: Sorry, that was an accidental commit. I've put it back to the default that it should have been. This should fix the concourse tests too.

**Pull request merged**